pkg/keystorage/keystorage.go

```diff
@@ -0,0 +1,193 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package keystorage provides the key storage implementation.
+package keystorage
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/subtle"
+	"fmt"
+	"sync"
+
+	"github.com/ProtonMail/gopenpgp/v2/helper"
+
+	"github.com/cosi-project/runtime/api/key_storage"
+)
+
+// KeyStorage is a key storage that can be used to store and retrieve the master key.
+//
+//nolint:govet
+type KeyStorage struct {
+	mx         sync.Mutex
+	underlying key_storage.Storage
+}
+
+// SetMasterKey sets the master key for the key storage, encrypts it using public key and stores it in slot id.
+func (ks *KeyStorage) SetMasterKey(masterKey []byte, slotID, slotPublicKey string) error {
```

Comment:
suggestion: we could call this `Initialize()` (or something similar) and make this function generate the random `masterKey` itself

Reply:
Renamed to `Initialize` and added a new method `InitializeRnd`.